### DISCUSSION TITLE
Integrate Shiny in an Existing Prism App

### DISCUSSION THREAD

**Discussion by vppetrov** [shinyorg/shiny#857] [Q&A] [answered]
We have an existing Prism app and we are trying to integrate Shiny in it. We are taking sample code from the Jobs sample. After adding the startup code and the calls to assembly: Shiny.ShinyApplication we start getting multiple errors:

```
Error	CS1061	'MainActivity' does not contain a definition for 'ShinyOnNewIntent' and no accessible extension method 'ShinyOnNewIntent' accepting a first argument of type 'MainActivity' could be found (are you missing a using directive or an assembly reference?)	MyPrismApp.Android	C:\Source\MyPrismApp.Android\Shiny.Generators\Shiny.Generators.AndroidActivitySourceGenerator\MainActivity.cs

Error	CS1061	'MainActivity' does not contain a definition for 'ShinyOnActivityResult' and no accessible extension method 'ShinyOnActivityResult' accepting a first argument of type 'MainActivity' could be found (are you missing a using directive or an assembly reference?)	MyPrismApp.Android	C:\Source\MyPrismApp.Android\Shiny.Generators\Shiny.Generators.AndroidActivitySourceGenerator\MainActivity.cs

Error	CS1061	'SplashActivity' does not contain a definition for 'ShinyOnCreate' and no accessible extension method 'ShinyOnCreate' accepting a first argument of type 'SplashActivity' could be found (are you missing a using directive or an assembly reference?)	MyPrismApp.Android	C:\Source\MyPrismApp.Android\Shiny.Generators\Shiny.Generators.AndroidActivitySourceGenerator\SplashActivity.cs

Error	CS1061	'SplashActivity' does not contain a definition for 'ShinyOnNewIntent' and no accessible extension method 'ShinyOnNewIntent' accepting a first argument of type 'SplashActivity' could be found (are you missing a using directive or an assembly reference?)	MyPrismApp.Android	C:\Source\MyPrismApp.Android\Shiny.Generators\Shiny.Generators.AndroidActivitySourceGenerator\SplashActivity.cs

Error	CS1061	'SplashActivity' does not contain a definition for 'ShinyOnActivityResult' and no accessible extension method 'ShinyOnActivityResult' accepting a first argument of type 'SplashActivity' could be found (are you missing a using directive or an assembly reference?)	MyPrismApp.Android	C:\Source\MyPrismApp.Android\Shiny.Generators\Shiny.Generators.AndroidActivitySourceGenerator\SplashActivity.cs

Error	CS1061	'SplashActivity' does not contain a definition for 'ShinyOnRequestPermissionsResult' and no accessible extension method 'ShinyOnRequestPermissionsResult' accepting a first argument of type 'SplashActivity' could be found (are you missing a using directive or an assembly reference?)	MyPrismApp.Android	C:\Source\MyPrismApp.Android\Shiny.Generators\Shiny.Generators.AndroidActivitySourceGenerator\SplashActivity.cs

Error		OnRequestPermissionsResult already exists on 'MyPrismApp.Droid.MainActivity', make sure you call the this.ShinyOnPermissionsResult hook for this	MyPrismApp.Android	C:\Source\MyPrismApp.Android\CSC		
Error		OnCreate already exists on 'MyPrismApp.Droid.MainActivity', make sure you call the this.ShinyOnCreate hook for this	MyPrismApp.Android	C:\Source\MyPrismApp.Android\CSC		
Error		FinishedLaunching already exists on your appdelegate.  Make sure to call the this.ShinyFinishedLaunching(new YourStartup());	MyPrismApp.iOS	C:\Source\MyPrismApp.iOS\AppDelegate.cs
```

It looks like Shiny is trying to override methods that we already override/use and it is trying to call some non-existing methods.

How can this be solved or is there a guide how to integrate Shiny in an existing Prism app that already has code in some methods?

UPDATE:

I figured out that Shiny creates partial classes under References -> Analyzers -> Shiny.Generators. These already contain some of the overrides, e.g. OnCreated, which then call the methods reported as missing, e.g. ShinyOnCreated. The issues are:
* I cannot override the same methods and thus cannot have custom code in them. That's quite a pity and almost a no-go, so I might need to go to the old way of registering the boilerplate myself.
* The methods like ShinyOnCreated should be available as they are in the library Shiny.Core and classes like AndroidHooks. However, these are not found, even if I explicitly add a reference to this project. Interestingly, the Jobs sample from Shiny can find them without a problem.

I still cannot compile anything as the methods are missing, so even the old boilerplate does not work..

**Comment by aritchie** [comment, marked as answer]
> It looks like Shiny is trying to override methods that we already override/use and it is trying to call some non-existing methods.

They do exist.  You have to look at the nuget package.  It targets monoandroid11+.  You need to adjust your targets for them to be picked up

> How can this be solved or is there a guide how to integrate Shiny in an existing Prism app that already has code in some methods?

There are two examples.  The first without codegen and the second for how to integrate with Prism.  Please make sure you look here in the future before sending questions
https://github.com/shinyorg/samples/tree/main/Boilerplate-NoCodeGen
https://github.com/shinyorg/samples/tree/main/Integration-Prism

**Reply by vppetrov** [reply]
Hi Allan and thanks for taking the time to answer.

> They do exist. You have to look at the nuget package. It targets monoandroid11+. You need to adjust your targets for them to be picked up

I just figured that out, changed from 10 to 11, now it finds them.

> There are two examples. The first without codegen and the second for how to integrate with Prism. Please make sure you look here in the future before sending questions.

I hadn't found them (searching in google for Prism + Shiny did not give me these results). I will review.

Could you explain how the following works (it's not Prism specific but Shiny in general):
E.g. on Android: I have my MainActivity partial class. Your code generator also create a partial MainActivity class which overrides the method OnResume(). If I want to have my own code inside OnResume() and I add it to my part of the partial class, your OnResume() override disappears. So how can I have both your implementation plus mine and make sure both are called?

**Comment by vppetrov** [comment]
Returning to the question about Prism: The sample at https://github.com/shinyorg/samples/tree/main/Integration-Prism uses a Prism app with a parameterless constructor. The class PrismApplication contains a second constructor with a parameter of type IPlatformInitializer (and even another one with two parameters). Is it possible to make Shiny's generated code call that specific constructor?

**Reply by aritchie** [reply]
No.  The Shiny generator doesn't look at Prism.
